blockly-rtc/src/MarkerUpdate.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Object for communicating an update to a Marker's location.
+ * @author navil@google.com (Navil Perez)
+ */
+
+import * as Blockly from 'blockly/dist';
+
+export default class MarkerUpdate {
+  constructor(id, type, blockId, fieldName) {
+    this.id = id;
+    this.type_ = type;
+    this.blockId_ = blockId;
+    this.fieldName_ = fieldName;
+  };
+
+  /**
+   * Create a MarkerUpdate from an event. Currently supports MarkerUpdates
+   * on blocks from a 'selected' UI event.
+   * @param {!Blockly.Event} event The event that creates a MarkerUpdate.
+   * @return {!MarkerUpdate} The MarkerUpdate representative of the event.
+   * @public
+   */  
+  static fromEvent(event) {
+    // TODO: Add support for a field marker on a change event.
+    const id = event.workspaceId;
+    const type = 'BLOCK';
+    const blockId = event.newValue;
+    const fieldName = null;
+    return new MarkerUpdate(id, type, blockId, fieldName);  
+  };
+
+  /**
+   * Decode the JSON into a MarkerUpdate.
+   * @param {!Object} json The JSON representation of the markerUpdate.
+   * @return {!MarkerUpdate} The MarkerUpdate represented by the JSON.
+   * @public
+   */
+  static fromJson(json) {
+    const markerLocation = json.markerLocation;
+    return new MarkerUpdate(
+        json.id,
+        markerLocation.type,
+        markerLocation.blockId,
+        markerLocation.fieldName);
+  };
+
+  /**
+   * Encode the MarkerUpdate as JSON.
+   * @return {!Object} The JSON representation of the MarkerUpdate.
+   * @public
+   */
+  toJson() {
+    return {
+      id: this.id,
+      markerLocation: this.getMarkerLocation()
+    };
+  };
+
+  /**
+   * Check if the combination of MarkerLocation properties describe a
+   * viable location.
+   * @return {!Boolean} Whether the MarkerUpdate has a viable location.
+   * @public
+   */  
+  hasLocation() {
+    if (this.type_ == 'FIELD' || this.blockId_ || this.fieldName_) {
```

Comment:
Should these be && instead?  It doesn't make sense to have || this.blockID_ in two of these cases.

Reply:
Yes, thank you for pointing that out.